### ISSUE TITLE
Relay style connections example

### ISSUE THREAD

**Issue by codepunkt** [prisma/nexus#22]
Very interesting library, thanks to everyone involved!

I'd be interested in a relay-style connection example! How'd you do these?

**Comment by tgriesser** [comment]
Will likely get something core-built in for this relatively soon, but just to demonstrate how you can go about building an abstraction like this for yourself (api will likely be refined):

```ts
export const User = objectType({
  name: 'User',
  definition(t) {
     t.id('id')
     // this is a connection, in the future it could be t.connection(...)
     t.field('friends', { 
       args: cursorArgs, 
       type: 'UserConnection', 
       resolve: () => // ... code to fetch connection shape
     })
  }
})

export const UserConnection = connectionType({type: 'User'})
```

Where `connectionType` and `cursorArgs` would be defined as:

```ts
import { objectType, intArg, stringArg, core } from "nexus";

interface RelayCursorConfig<T extends core.GetGen<"objectNames">> {
  type: T;
  edgeDefinition?: (t: core.ObjectDefinitionBlock<string>) => any;
  connectionDefinition?: (t: core.ObjectDefinitionBlock<string>) => any;
  pageInfoDefinition?: (t: core.ObjectDefinitionBlock<string>) => any;
}

export const cursorArgs = {
  first: intArg({
    description:
      "Forward pagination, returns first N edges based on after cursor",
  }),
  after: stringArg({
    description: "Takes the cursor type.",
  }),
};

export const backwardArgs = {
  last: intArg({
    description: "Backward pagination, takes a non-negative integer",
  }),
  before: stringArg({
    description: "Takes the cursor type.",
  }),
};

/**
 * Helper for creating a relay-style cursor.
 * @param type
 */
export function connectionType<T extends core.GetGen<"objectNames">>(
  config: RelayCursorConfig<T>
) {
  const {
    type,
    edgeDefinition,
    connectionDefinition,
    pageInfoDefinition,
  } = config;
  const Connection = objectType({
    name: `${type}Connection`,
    definition(t) {
      t.list.field("edges", {
        type: Edge,
      });
      t.field("pageInfo", { type: PageInfo });
      if (typeof connectionDefinition === "function") {
        connectionDefinition(t);
      }
    },
  });

  const Edge = objectType({
    name: `${type}Edge`,
    definition(t) {
      t.string('cursor') // TODO: add cursor resolver
      t.field("node", { type });
      if (typeof edgeDefinition === "function") {
        edgeDefinition(t);
      }
    },
  });
  const PageInfo = pageInfoDefinition
    ? objectType({
        name: `${type}PageInfo`,
        definition(t) {
          t.boolean("hasPreviousPage");
          t.boolean("hasNextPage");
          pageInfoDefinition(t);
        },
      })
    : DefaultPageInfo;
  return [Connection, Edge, PageInfo];
}

const DefaultPageInfo = objectType({
  name: `PageInfo`,
  definition(t) {
    t.boolean("hasPreviousPage");
    t.boolean("hasNextPage");
  },
});
```
There's also a way we can do this a little more intelligently using some internal builder APIs so you could use the `UserConnection` object directly rather than the string. This would require specifying the type name of the connection though, in order to keep type-safety when it's used:

```ts
export const User = objectType({
  name: 'User',
  definition(t) {
     t.id('id')
     t.field('friends', { 
       args: cursorArgs, 
       type: UserConnection, // Note that this isn't a string anymore
       resolve: () => // ... code to fetch connection shape
     })
  }
})

export const UserConnection = connectionType({ 
  type: 'User', 
  connectionName: 'UserConnection' 
})
```

implementation:

```ts
interface RelayCursorConfig<
  T extends core.GetGen<"objectNames">,
  ConnectionName extends string
> {
  type: T;
  connectionName: ConnectionName;
  edgeDefinition?: (t: core.ObjectDefinitionBlock<string>) => any;
  connectionDefinition?: (t: core.ObjectDefinitionBlock<string>) => any;
  pageInfoDefinition?: (t: core.ObjectDefinitionBlock<string>) => any;
}

/**
 * Helper for creating a relay-style cursor.
 * @param type
 */
export function makeRelayCursor<
  T extends core.GetGen<"objectNames">,
  ConnectionName extends string
>(config: RelayCursorConfig<T, ConnectionName>) {
  return core.nexusWrappedType(config.connectionName, (builder) => {
    const {
      type,
      edgeDefinition,
      connectionDefinition,
      pageInfoDefinition,
    } = config;

    const Edge = objectType({
      name: `${type}Edge`,
      definition(t) {
        t.string('cursor') // TODO: add cursor resolver
        t.field("node", { type });
        if (typeof edgeDefinition === "function") {
          edgeDefinition(t);
        }
      },
    });

    const PageInfo = pageInfoDefinition
      ? objectType({
          name: `${type}PageInfo`,
          definition(t) {
            t.boolean("hasPreviousPage");
            t.boolean("hasNextPage");
            pageInfoDefinition(t);
          },
        })
      : DefaultPageInfo;

    builder.addType(Edge);
    builder.addType(PageInfo);

    return objectType({
      name: config.connectionName,
      definition(t: core.ObjectDefinitionBlock<string>) {
        t.list.field("edges", {
          type: Edge,
        });
        t.field("pageInfo", { type: PageInfo });
        if (typeof connectionDefinition === "function") {
          connectionDefinition(t);
        }
      },
    });
  });
}

const DefaultPageInfo = objectType({
  name: `PageInfo`,
  definition(t) {
    t.boolean("hasPreviousPage");
    t.boolean("hasNextPage");
  },
});
```

**Comment by tgriesser** [comment]
#143 is looking to have a way to standardize this - checkout the code there and the kitchen sink example, available under the newly published `0.12.0-beta.1`.
